compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -4,1440 +4,376 @@
-  return resolveWithPossibleConflict(rhs->vectorLayout, opOperand);
-}
-
-ChangeResult DistributionLayout::resolve(const VectorLayoutInterface &rhs,
-                                         bool force) {
-  // If forced, set the layout regardless of a possible conflict.
-  if (force) {
-    bool changed = (vectorLayout != rhs);
-    setInnerLayout(rhs);
-    return changed ? ChangeResult::Change : ChangeResult::NoChange;
-  }
-
-  ResolutionResult result = doResolution(rhs);
-
-  switch (result) {
-  case ResolutionResult::NoChange:
-    return ChangeResult::NoChange;
-  case ResolutionResult::Change:
-    return ChangeResult::Change;
-  case ResolutionResult::Conflict: {
-    llvm::errs() << "Layout conflict at: " << *this << "\n";
-    llvm::errs() << "With: " << rhs << "\n";
-    llvm::report_fatal_error("Layout conflict should have been handled with "
-                             "resolveWithPossibleConflict instead");
-  }
-  }
-
-  // This return will never be reached, but it's here to make the compiler
-  // happy.
-  return ChangeResult::NoChange;
-}
-
-ChangeResult DistributionLayout::resolve(const DistributionLayout *rhs,
-                                         bool force) {
-  assert(rhs && "layout to resolve with should not be null");
-  return resolve(rhs->vectorLayout, force);
-}
-
-void DistributionLayout::print(raw_ostream &os) const {
-  if (vectorLayout) {
-    os << " " << vectorLayout;
-  } else {
-    os << "Uninitialized";
-  }
-}
-
-void DistributionLayout::onUpdate(DataFlowSolver *solver) const {
-  AnalysisState::onUpdate(solver);
+void LayoutInfo::propagateLayoutForward(Value val) {
+  LDBG() << "Propagating layout forward for value: " << val << "\n";
+  VectorLayoutInterface layout = getLayout(val);
+  for (OpOperand &use : val.getUses()) {
+    unsigned operandIdx = use.getOperandNumber();
+    Operation *user = use.getOwner();
+
+    if (auto forOp = dyn_cast<scf::ForOp>(user)) {
+      Value arg = forOp.getTiedLoopRegionIterArg(&use);
+      Value result = forOp.getTiedLoopResult(&use);
+      setLayoutIfUnset(arg, layout);
+      setLayoutIfUnset(result, layout);
+      continue;
+    }
 
-  Value value = cast<Value>(anchor);
+    if (auto yieldOp = dyn_cast<scf::YieldOp>(user)) {
+      Operation *parentOp = yieldOp->getParentOp();
+      if (auto forOp = dyn_cast<scf::ForOp>(parentOp)) {
+        Value arg = forOp.getRegionIterArg(operandIdx);
+        Value result = forOp->getResult(operandIdx);
+        setLayoutIfUnset(arg, layout);
+        setLayoutIfUnset(result, layout);
+        continue;
+      }
+      if (auto ifOp = dyn_cast<scf::IfOp>(parentOp)) {
+        Value thenArg = ifOp.getThenRegion().getArgument(operandIdx);
+        Value elseArg = ifOp.getElseRegion().getArgument(operandIdx);
+        Value result = ifOp->getResult(operandIdx);
+        setLayoutIfUnset(thenArg, layout);
+        setLayoutIfUnset(elseArg, layout);
+        setLayoutIfUnset(result, layout);
+        continue;
+      }
```

Comment:
```suggestion
      }
      assert(false && "must add logic for this yield case");
```

Or some other way of making it easier to debug if we get an scf.while or something one day. I guess we want to evenutally better encapsulate the op specific logic in an interface or something.

Reply:
The analysis is "best case" I don't think adding an assert there is a good idea.